src/shmem_team.c

```diff
@@ -0,0 +1,455 @@
+        }
+    }
+
+    size_t psync = shmem_internal_team_choose_psync(parent_team, SYNC);
+
+    shmem_internal_barrier(parent_team->start, parent_team->stride, parent_team->size,
+                           &shmem_internal_psync_barrier_pool[psync]);
+
+    shmem_internal_team_release_psyncs(parent_team, SYNC);
+
+    if (myteam->psync_idx == -1)
+        return -1;
+    else
+        return 0;
+}
+
+int shmem_internal_team_split_2d(shmem_internal_team_t *parent_team, int xrange,
+                                 const shmemx_team_config_t *xaxis_config, long xaxis_mask,
+                                 shmem_internal_team_t **xaxis_team, const shmemx_team_config_t *yaxis_config,
+                                 long yaxis_mask, shmem_internal_team_t **yaxis_team)
+{
+    const int parent_start = parent_team->start;
+    const int parent_stride = parent_team->stride;
+    const int parent_size = parent_team->size;
+    const int num_xteams = ceil( parent_size / (float)xrange );
+    const int num_yteams = xrange;
+
+    int start = parent_start;
+    int ret = 0;
+
+    for (int i = 0; i < num_xteams; i++) {
+	int xsize = (i == num_xteams - 1 && parent_size % xrange) ? parent_size % xrange : xrange;
+
+        if (shmem_internal_pe_in_active_set(shmem_internal_my_pe, start, parent_stride, xsize) != -1) {
+            ret = shmem_internal_team_split_strided(parent_team, start, parent_stride,
+                                            xsize, xaxis_config, xaxis_mask, xaxis_team);
+            if (ret) {
+                RAISE_ERROR_STR("x-axis 2D strided split failed");
+            }
+        }
+        start += xrange * parent_stride;
+    }
+
+    start = parent_start;
+
+    for (int i = 0; i < num_yteams; i++) {
+        int remainder = parent_size % xrange;
+        int yrange = parent_size / xrange;
+        int ysize = (remainder && i < remainder) ? yrange + 1 : yrange;
+
+        if (shmem_internal_pe_in_active_set(shmem_internal_my_pe, start, xrange*parent_stride, ysize) != -1) {
+            ret = shmem_internal_team_split_strided(parent_team, start, xrange*parent_stride,
+                                            ysize, yaxis_config, yaxis_mask, yaxis_team);
+            if (ret) {
+                RAISE_ERROR_STR("y-axis 2D strided split failed");
+            }
+        }
+        start += parent_stride;
+    }
+
+    size_t psync = shmem_internal_team_choose_psync(parent_team, SYNC);
+
+    shmem_internal_barrier(parent_start, parent_stride, parent_size,
+                           &shmem_internal_psync_barrier_pool[psync]);
+
+    shmem_internal_team_release_psyncs(parent_team, SYNC);
+
+    return 0;
+}
+
+int shmem_internal_team_destroy(shmem_internal_team_t *team)
+{
+
+    if (team == SHMEMX_TEAM_INVALID || team == &shmem_internal_team_world ||
+        team == &shmem_internal_team_shared) {
+        return -1;
+    } else if ((psync_pool_avail[team->psync_idx / CHAR_BIT] >> team->psync_idx) & 1) {
```

Comment:
Please move the above expression to a query function: `shmem_internal_bit_fetch`